TVShows/TVShows/Modules/ShowDetails/ShowDetailsPresenter.swift

```diff
@@ -0,0 +1,103 @@
+//
+//  ShowDetailsPresenter.swift
+//  TVShows
+//
+//  Created by Ivan Milicevic on 22/10/2018.
+//  Copyright (c) 2018 Ivan Milicevic. All rights reserved.
+//
+//  This file was generated by the 🐍 VIPER generator
+//
+
+import UIKit
+
+final class ShowDetailsPresenter {
+
+    // MARK: - Private properties -
+
+    private unowned var _view: ShowDetailsViewInterface
+    private var _interactor: ShowDetailsInteractorInterface
+    private var _wireframe: ShowDetailsWireframeInterface
+
+    private var _show: Show
+    private var _showDetails: ShowDetails?
+    private var _episodesArray: [ShowEpisode] = []
+
+    // MARK: - Lifecycle
+
+    init(wireframe: ShowDetailsWireframeInterface, view: ShowDetailsViewInterface, interactor: ShowDetailsInteractorInterface, show: Show) {
+        _wireframe = wireframe
+        _view = view
+        _interactor = interactor
+        _show = show
+    }
+
+}
+
+// MARK: - Extensions -
+
+extension ShowDetailsPresenter: ShowDetailsPresenterInterface {
+
+    func reloadData() {
+        fetchData()
+    }
+
+    func getEpisodesCount() -> Int {
+        return _episodesArray.count
+    }
+
+    func getShowDetails() -> ShowDetails? {
+        return _showDetails
+    }
+
+    func numberOfRowsInSection() -> Int {
+        return _episodesArray.count
+    }
+
+    func getEpisode(at index: Int) -> ShowEpisode {
+        return _episodesArray[index]
+    }
+
+    func removeEpisode(at index: Int) {
+        _episodesArray.remove(at: index)
+    }
+
+    func didSelectEpisode(at index: Int) {
+        if index == 0 || index == 1 {
+            return
+        }
+        _wireframe.navigate(to: .episodeDetails(episode: _episodesArray[index-2]))
+    }
+
+    func didPressAddNewEpisodeButton() {
+        _wireframe.navigate(to: .addNewEpisode(show: _show))
+    }
+
+    func viewDidLoad() {
+        fetchData()
+    }
+
+    func didPressBackButton() {
+        _wireframe.navigate(to: .home)
+    }
+
+    private func fetchData() {
+        _interactor.fetchShowDetails(show: _show) { result in
```

Comment:
You are strong capturing self inside closure! Never do that because you can easily create memory leak or retain something longer than needed and possibly cause crash.

Current example is very dangerous. So, lets break it in components. First of all, let say you have shitty internet connection. You have opened show details screen and it takes forever to load it. You are annoyed with waiting and you press back. So API call did not finish!

Next thing, to notice is that presenter holds weak/unowned reference to view, so when view is dealloacted, it won't automatically dealloc presenter since you have retained self in API call closure (API which didn't finished yet).

And, when api call finishes it will return to this closure. But now you don't have view anymore since it is deallocated (you pressed back) and accessing unowned variable which is deallocated will cause crash!

So to avoid that always use `[weak self]` and guard letting since it stop execution immediately. Something like this:

```swift
_interactor.fetchShowDetails(show: _show) { [weak self] result in 
    guard let self = self else { return }
    // Some mumbo jumbo where you use self...
}
```